src/Features/CSharp/Portable/ConvertLinq/ConvertForEachToLinqQuery/AbstractConverter.cs

```diff
@@ -93,5 +114,158 @@ private static QueryClauseSyntax CreateQueryClause(ExtendedSyntaxNode node)
+            ForEachStatementSyntax forEachStatement,
+            ExpressionSyntax receiverForInvocation,
+            IEnumerable<SyntaxTrivia> leadingCommentsTrivia,
+            IEnumerable<SyntaxTrivia> trailingCommentsTrivia,
+            ExpressionSyntax selectExpression,
+            ref int currentExtendedNodeIndex)
+        {
+            leadingCommentsTrivia = forEachStatement.ForEachKeyword.GetAllTrivia().Concat(leadingCommentsTrivia);
+
+            // Recursively create linq invocations, possibly updating the receiver (Where invocations), to get the inner expression for
+            // the lambda body for the linq invocation to be created for this foreach statement. For example:
+            //
+            // INPUT:
+            //   foreach (var n1 in c1)
+            //      foreach (var n2 in c2)
+            //          if (n1 > n2)
+            //              yield return n1 + n2;
+            //
+            // OUTPUT:
+            //   c1.SelectMany(n1 => c2.Where(n2 => n1 > n2).Select(n2 => n1 + n2))
+            //
+            bool hasForEachChild = false;
+            var lambdaBody = CreateLinqInvocationForExtendedNode(selectExpression, ref currentExtendedNodeIndex, ref receiverForInvocation, ref hasForEachChild);
+            var lambda = SyntaxFactory.SimpleLambdaExpression(
+                SyntaxFactory.Parameter(
+                    forEachStatement.Identifier.WithPrependedLeadingTrivia(
+                    SyntaxNodeOrTokenExtensions.GetTrivia(forEachStatement.Type.GetFirstToken())
+                        .FilterComments(addElasticMarker: false))),
+                lambdaBody)
+                .WithCommentsFrom(leadingCommentsTrivia, trailingCommentsTrivia,
+                    forEachStatement.OpenParenToken, forEachStatement.InKeyword, forEachStatement.CloseParenToken);
+
+            // Create Select or SelectMany linq invocation for this foreach statement. For example:
+            //
+            // INPUT:
+            //   foreach (var n1 in c1)
+            //      ...
+            //
+            // OUTPUT:
+            //   c1.Select(n1 => ...
+            //      OR
+            //   c1.SelectMany(n1 => ...
+            //
+            var invokedMethodName = !hasForEachChild ? "Select" : "SelectMany";
+            return SyntaxFactory.InvocationExpression(
+                SyntaxFactory.MemberAccessExpression(
+                    SyntaxKind.SimpleMemberAccessExpression,
+                    receiverForInvocation.Parenthesize(),
+                    SyntaxFactory.IdentifierName(invokedMethodName)),
+                SyntaxFactory.ArgumentList(SyntaxFactory.SingletonSeparatedList(
+                    SyntaxFactory.Argument(lambda))));
+        }
+
+        private ExpressionSyntax CreateLinqInvocationForExtendedNode(
+            ExpressionSyntax selectExpression,
+            ref int extendedNodeIndex,
+            ref ExpressionSyntax receiver,
+            ref bool hasForEachChild)
```

Comment:
can we doc what all these ref params are and how they change.  it was not clear to me

Reply:
Sure.